mk/Common.mk

```diff
@@ -231,123 +261,157 @@ include $(CFG_FLAGS_MAKE)
-RISCV_PREFIX     = riscv32-$(GNU_VENDOR)-elf-
+# The so-called "COREV Environment Variables".
+#
+# First, check that the COREV environment variables have been defined.
+ifndef CV_SW_TOOLCHAIN
+$(error Must define CV_SW_TOOLCHAIN)
+endif
+ifndef CV_SW_PREFIX
+$(error Must define CV_SW_PREFIX)
+endif
+ifndef CV_SW_MARCH
+$(error Must define CV_SW_MARCH)
+endif
+#ifndef CV_SW_CFLAGS
+#$(warning CV_SW_CFLAGS is optional)
+#endif
+
+# Use these to set defaults for toolchain vars. A few typical examples...
+#RISCV            = /opt/riscv
+#RISCV_PREFIX     = riscv32-unknown-elf-
+#RISCV_EXE_PREFIX = $(RISCV)/bin/$(RISCV_PREFIX)
+#RISCV_CFLAGS     =
+RISCV            = $(CV_SW_TOOLCHAIN)
+RISCV_PREFIX     = $(CV_SW_PREFIX)
 RISCV_EXE_PREFIX = $(RISCV)/bin/$(RISCV_PREFIX)
-RISCV_CC         = $(GNU_CC)
-RISCV_MARCH      = $(call RESOLVE_FLAG2,$(TEST_GNU_MARCH),$(GNU_MARCH))
-RISCV_CFLAGS     = $(call RESOLVE_FLAG2,$(TEST_GNU_CFLAGS),$(GNU_CFLAGS))
-endif
 
-ifeq ($(COREV_YES),YES)
-ifeq ($(call IS_YES,$(TEST_COREV_NOT_SUPPORTED)),YES)
-$(error test [$(TEST)] does not support the COREV toolchain)
-endif
-RISCV            = $(COREV_SW_TOOLCHAIN)
-RISCV_PREFIX     = riscv32-$(COREV_VENDOR)-elf-
-RISCV_EXE_PREFIX = $(RISCV)/bin/$(RISCV_PREFIX)
-RISCV_CC         = $(COREV_CC)
-RISCV_MARCH      = $(call RESOLVE_FLAG2,$(TEST_COREV_MARCH),$(COREV_MARCH))
-RISCV_CFLAGS     = $(call RESOLVE_FLAG2,$(TEST_COREV_CFLAGS),$(COREV_CFLAGS))
-endif
+RISCV_CC         = gcc
+RISCV_MARCH      = rv32imc
 
-ifeq ($(PULP_YES),YES)
-ifeq ($(call IS_YES,$(TEST_PULP_NOT_SUPPORTED)),YES)
-$(error test [$(TEST)] does not support the PULP toolchain)
-endif
-RISCV            = $(PULP_SW_TOOLCHAIN)
-RISCV_PREFIX     = riscv32-$(PULP_VENDOR)-elf-
-RISCV_EXE_PREFIX = $(RISCV)/bin/$(RISCV_PREFIX)
-RISCV_CC         = $(PULP_CC)
-RISCV_MARCH      = $(call RESOLVE_FLAG2,$(TEST_PULP_MARCH),$(PULP_MARCH))
-RISCV_CFLAGS     = $(call RESOLVE_FLAG2,$(TEST_PULP_CFLAGS),$(PULP_CFLAGS))
-endif
+RISCV_MARCH      = $(call RESOLVE_FLAG2,$(TEST_RISCV_MARCH),$(COREV_SW_MARCH))
+RISCV_MARCH      = $(call RESOLVE_FLAG2,$(CFG_COREV_MARCH),$(TEST_RISCV_MARCH))
 
-ifeq ($(LLVM_YES),YES)
-ifeq ($(call IS_YES,$(TEST_LLVM_NOT_SUPPORTED)),YES)
-$(error test [$(TEST)] does not support the LLVM toolchain)
-endif
-RISCV            = $(LLVM_SW_TOOLCHAIN)
-RISCV_PREFIX     = riscv32-$(LLVM_VENDOR)-elf-
-RISCV_EXE_PREFIX = $(RISCV)/bin/$(RISCV_PREFIX)
-RISCV_CC         = $(LLVM_CC)
-RISCV_MARCH      = $(call RESOLVE_FLAG2,$(TEST_LLVM_MARCH),$(LLVM_MARCH))
-RISCV_CFLAGS     = $(call RESOLVE_FLAG2,$(TEST_LLVM_CFLAGS),$(LLVM_CFLAGS))
-RISCV_CFLAGS    += -menable-experimental-extensions
-endif
+RISCV_CFLAGS     = $(call RESOLVE_FLAG2,$(CFG_RISCV_CFLAGS),$(TEST_RISCV_CFLAGS))
+RISCV_CFLAGS     = $(call RESOLVE_FLAG2,$(TEST_RISCV_CFLAGS),$(COREV_SW_CFLAGS))
 
-CFLAGS ?= -Os -g -static -mabi=ilp32 -march=$(RISCV_MARCH) -Wall -pedantic
+CFLAGS ?= -Os -g -static -mabi=ilp32 -march=$(RISCV_MARCH) -Wall -pedantic $(RISCV_CFLAGS)
```

Comment:
for the future, we may want to parameterize -mabi as well (e.g. for floating point)

Reply:
To elaborate a bit further, I think when we allow partial changes to the CFLAGS like we do with these parameters, we may want to break it up into sections s.t. we can modify individual parts, without having to provide a full CFLAGS-argument list. I recon that for example -Os -g may be something we would want to tamper with at certain points

Reply:
I don't know what posix dictates should happen when one option is repeated with a different value. Can the CFLAGS variable just overrule all previous arguments?

The gcc man page says this about "-O" options:
> If you use multiple -O options, with or without level numbers, the last such option is the one that is effective.

Reply:
> The gcc man page says this about "-O" options:
> 
> > If you use multiple -O options, with or without level numbers, the last such option is the one that is effective.

This may be compiler specific - do you know how LLVM treats this situation?

Reply:
I couldn't find the equivalent of gcc's "the last such option is the one that is effective" in clang documentation.

Posix has a relevant guideline. "Guideline 11" here:
https://pubs.opengroup.org/onlinepubs/9699919799.2018edition/basedefs/V1_chap12.html#tag_12_02
_"If an option that has option-arguments is repeated, the option and option-argument combinations should be interpreted in the order specified on the command line."_

Reply:
Perhaps we could define (yet) another var, CV_SW_CFLAGS and use what we have now for the default.

Reply:
We now have `CV_SW_CFLAGS` as an optional COREV environment variable.  See `mk/README.md` for more info.